qiskit/quantum_info/synthesis/one_qubit_decompose.py

```diff
@@ -287,28 +287,30 @@ def _circuit_zyz(theta, phi, lam, phase, simplify=True, atol=DEFAULT_ATOL):
         circuit = QuantumCircuit(qr)
         if not simplify:
             atol = -1.0
-        if abs(theta) < atol:
-            tot = _mod_2pi(phi + lam, atol)
-            if abs(tot) > atol:
-                circuit._append(RZGate(tot), [qr[0]], [])
-                gphase += tot / 2
+        if abs(_mod_2pi(theta)) < atol:
+            lam, phi = lam + phi, 0
+            if abs(_mod_2pi(lam)) > atol:
+                circuit._append(RZGate(lam), [qr[0]], [])
+                gphase += lam / 2
             circuit.global_phase = gphase
             return circuit
-        if abs(theta - np.pi) < atol:
+        if abs(_mod_2pi(theta - np.pi)) < atol:
             gphase += phi
             lam, phi = lam - phi, 0
-        lam = _mod_2pi(lam, atol)
         if abs(_mod_2pi(lam + np.pi)) < atol:
             lam, theta, phi = 0, -theta, lam + phi
-        if abs(lam) > atol:
+        if abs(_mod_2pi(phi + np.pi)) < atol:
```

Comment:
maybe better to make it a single if:
```python
if abs(_mod_2pi(lam-np.pi))<atol or abs(_mod_2pi(phi-np.pi))<atol:
   theta, phi, lam = -theta, phi+pi, lam+pi
```
Then we could also make the if a bit smarter to cover different concepts of optimality:
eg add a flag `allow_non_canonical` (default `True`) like we have `simplify`, and the check can be:
```python
if allow_non_canonical and (abs(_mod_2pi(lam-np.pi))<atol or abs(_mod_2pi(phi-np.pi))<atol):
   theta, phi, lam = -theta, phi+pi, lam+pi
```
which I think covers the 2 different behaviours one might desire from the decomposer.

Reply:
This reminds me that I haven't been consistent about guarding with `simplify`.

Reply:
Before I refactored it was wrapping "most" of the elisions. I refactored it to set atol=-1 and thus wrap all elisions